File: rs_holocubic/calendar_data.py
"""只读 CalendarTask 当前子日历的当天日格，保留完成标记并生成中文分页。"""
import hashlib
import html
import json
import os
import re
from contextlib import closing
from datetime import date
from pathlib import Path
from PIL import Image, ImageDraw
from radar_data import read_db
import holo_usb_display as display

MAX_PAGES = 1


def decode_text(value):
    """还原 CalendarTask 的竖线实体包装，移除已观察到的字体颜色标签，不执行HTML。"""
    value = re.sub(r"\|(&[A-Za-z0-9#]+;)\|", lambda match: html.unescape(match.group(1)), value)
    value = re.sub(r"</?font\b[^>]*>", "", value, flags=re.IGNORECASE)
    return html.unescape(value).replace("\xa0", " ")


def parse_items(content):
    """每个非空源文本行对应一项；仅行首 [+] 表示完成，保留其余文本与顺序。"""
    items = []
    for line in decode_text(content).splitlines():
        text = line.strip()
        done = text.startswith("[+]")
        if done:
            text = text[3:].lstrip()
        if text:
            items.append({"text": text, "done": done})
    return items


def collect(db_path=None, day=None):
    """按本机日期读取当前子日历；只访问日格及选中日历信息，不读取账号凭据或历史。"""
    path = Path(db_path) if db_path else Path(os.environ["APPDATA"]) / "CalendarTask/Db/calendar.db"
    day = date.today() if day is None else day
    with closing(read_db(path)) as db:
        db.execute("BEGIN")
        settings = {row["st_name"]: row for row in db.execute(
            "SELECT st_name,st_nval,st_sval FROM setting_table WHERE st_name IN ('sys_current_sub_account','sys_sub_account_list','group_id')")}
        account_id = settings["sys_current_sub_account"]["st_nval"]
        if type(account_id) is not int or account_id < 0:
            raise ValueError("请在日历清单中选择一个具体子日历")
        accounts = json.loads(decode_text(settings["sys_sub_account_list"]["st_sval"]))["vdata"]["list"]
        account = next(row["name"] for row in accounts if row["id"] == account_id)
        group = settings["group_id"]["st_sval"]
        rows = db.execute("""SELECT it_content FROM item_table
            WHERE u_id=? AND pj_id=0 AND COALESCE(group_id,'')=? AND it_unique_id=?""",
            (account_id, group, "dkcal_mdays_"+day.strftime("%Y%m%d"))).fetchall()
    if len(rows) > 1:
        raise ValueError("当天日格存在多条记录，需先核对日历数据，未擅自合并")
    items = parse_items(rows[0]["it_content"]) if rows else []
    data = {"date": day.isoformat(), "weekday": "周"+"一二三四五六日"[day.weekday()],
            "account": account, "account_id": account_id, "items": items,
            "total": len(items), "done": sum(item["done"] for item in items), "layout": 2}
    data["signature"] = hashlib.sha256(json.dumps(data, ensure_ascii=False, sort_keys=True).encode("utf-8")).hexdigest()[:16]
    data["db_path"] = str(path)
    return data


def render(snapshot):
    """仅渲染紧凑的第一页，溢出明确提示；完成行保留删除线，不自动翻页。"""
    font = display.load_font(14)
    probe = ImageDraw.Draw(Image.new("RGB", (320, 240)))
    rows, y, truncated = [], 60, False
    for index, item in enumerate(snapshot["items"], 1):
        lines = display.wrap_text(probe, item["text"], font, 270)
        for line_index, line in enumerate(lines):
            if y+18 > 217:
                truncated = True
                break
            rows.append((index, line_index == 0 or not rows, line, item["done"], y))
            y += 18
        if truncated:
            break
        y += 6
    pages = [rows]
    output = []
    for page_index, page in enumerate(pages, 1):
        frame = Image.new("RGB", (320, 240), (0, 0, 0))
        draw = ImageDraw.Draw(frame)
        draw.text((12, 7), "日历清单", font=display.load_font(18, bold=True), fill=(222, 247, 255), anchor="lt")
        draw.text((226, 10), "完成 %d/%d" % (snapshot["done"], snapshot["total"]), font=display.load_font(11), fill=(81, 221, 199), anchor="lt")
        heading = snapshot["date"]+"  "+snapshot["weekday"]+" · "+snapshot["account"]
        heading_font = display.fit_single_line_font(draw, heading, 296, 11, 8)
        draw.text((12, 33), heading, font=heading_font, fill=(120, 155, 169), anchor="lt")
        draw.line((12, 51, 308, 51), fill=(30, 57, 65), width=1)
        if not page:
            draw.text((75, 116), "当日暂无任务", font=display.load_font(20), fill=(133, 166, 174), anchor="lt")
        for number, show_number, text, done, top in page:
            color = (121, 146, 154) if done else (227, 241, 247)
            if show_number:
                draw.text((12, top+2), "%02d" % number, font=display.load_font(10), fill=(81, 221, 199), anchor="lt")
            draw.text((38, top), text, font=font, fill=color, anchor="lt")
            if done:
                box = draw.textbbox((38, top), text, font=font, anchor="lt")
                draw.line((box[0], (box[1]+box[3])//2, box[2], (box[1]+box[3])//2), fill=color, width=1)
        draw.line((12, 220, 308, 220), fill=(30, 57, 65), width=1)
        draw.text((12, 225), "本地只读", font=display.load_font(10), fill=(120, 155, 169), anchor="lt")
        draw.text((158, 225), "仅第一页 · 还有内容" if truncated else "固定显示第一页", font=display.load_font(10), fill=(120, 155, 169), anchor="lt")
        output.append(display.encode_jpeg(frame))
    return output

File: rs_holocubic/src/calendar_data.rs
//! Read-only CalendarTask day-cell collect (no USB).

use crate::render::{self, RenderError};
use chrono::Local;
use html_escape::decode_html_entities;
use regex::Regex;
use rusqlite::{Connection, OpenFlags};
use serde_json::{json, Value};
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::OnceLock;

fn entity_wrap_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"\|(&[A-Za-z0-9#]+;)\|").unwrap())
}

fn font_tag_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"(?i)</?font\b[^>]*>").unwrap())
}

pub fn decode_text(value: &str) -> String {
    let step = entity_wrap_re().replace_all(value, |caps: &regex::Captures| {
        decode_html_entities(&caps[1]).to_string()
    });
    let step = font_tag_re().replace_all(&step, "");
    decode_html_entities(&step).replace(" ", " ")
}

pub fn parse_items(content: &str) -> Vec<Value> {
    let mut items = Vec::new();
    for line in decode_text(content).lines() {
        let mut text = line.trim().to_string();
        let mut done = false;
        if let Some(rest) = text.strip_prefix("[+]") {
            done = true;
            text = rest.trim_start().to_string();
        }
        if !text.is_empty() {
            items.push(json!({"text": text, "done": done}));
        }
    }
    items
}

fn open_ro(path: &Path) -> Result<Connection, String> {
    let uri = format!("file:{}?mode=ro", path.to_string_lossy().replace('\\', "/"));
    let db = Connection::open_with_flags(
        &uri,
        OpenFlags::SQLITE_OPEN_READ_ONLY | OpenFlags::SQLITE_OPEN_URI,
    )
    .map_err(|e| e.to_string())?;
    db.execute_batch("PRAGMA query_only=ON")
        .map_err(|e| e.to_string())?;
    Ok(db)
}

fn default_db_path() -> PathBuf {
    if let Ok(appdata) = std::env::var("APPDATA") {
        return PathBuf::from(appdata).join("CalendarTask/Db/calendar.db");
    }
    std::env::var_os("HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("."))
        .join("CalendarTask/Db/calendar.db")
}

/// Python `json.dumps(obj, ensure_ascii=False, sort_keys=True)` spacing.
fn json_dumps_sorted(value: &Value) -> String {
    match value {
        Value::Null => "null".into(),
        Value::Bool(b) => if *b { "true" } else { "false" }.into(),
        Value::Number(n) => n.to_string(),
        Value::String(s) => serde_json::to_string(s).unwrap_or_else(|_| "\"\"".into()),
        Value::Array(arr) => {
            let parts: Vec<String> = arr.iter().map(json_dumps_sorted).collect();
            format!("[{}]", parts.join(", "))
        }
        Value::Object(map) => {
            let mut keys: Vec<&String> = map.keys().collect();
            keys.sort();
            let parts: Vec<String> = keys
                .into_iter()
                .map(|k| {
                    format!(
                        "{}: {}",
                        serde_json::to_string(k).unwrap(),
                        json_dumps_sorted(&map[k])
                    )
                })
                .collect();
            format!("{{{}}}", parts.join(", "))
        }
    }
}

pub fn collect(db_path: Option<&Path>, day: Option<chrono::NaiveDate>) -> Result<Value, String> {
    let path = db_path.map(Path::to_path_buf).unwrap_or_else(default_db_path);
    let day = day.unwrap_or_else(|| Local::now().date_naive());
    let result = {
        let db = open_ro(&path)?;
        let mut settings: HashMap<String, (i64, String)> = HashMap::new();
        {
            let mut stmt = db
                .prepare(
                    "SELECT st_name,st_nval,st_sval FROM setting_table WHERE st_name IN ('sys_current_sub_account','sys_sub_account_list','group_id')",
                )
                .map_err(|e| e.to_string())?;
            let rows = stmt
                .query_map([], |row| {
                    Ok((
                        row.get::<_, String>(0)?,
                        row.get::<_, i64>(1)?,
                        row.get::<_, String>(2)?,
                    ))
                })
                .map_err(|e| e.to_string())?;
            for row in rows {
                let (name, nval, sval) = row.map_err(|e| e.to_string())?;
                settings.insert(name, (nval, sval));
            }
        }
        let account_id = settings
            .get("sys_current_sub_account")
            .map(|v| v.0)
            .ok_or_else(|| "请在日历清单中选择一个具体子日历".to_string())?;
        if account_id < 0 {
            return Err("请在日历清单中选择一个具体子日历".into());
        }
        let list_raw = settings
            .get("sys_sub_account_list")
            .map(|v| v.1.as_str())
            .ok_or_else(|| "请在日历清单中选择一个具体子日历".to_string())?;
        let accounts: Value =
            serde_json::from_str(&decode_text(list_raw)).map_err(|e| e.to_string())?;
        let list = accounts
            .pointer("/vdata/list")
            .and_then(|v| v.as_array())
            .ok_or_else(|| "子日历列表无效".to_string())?;
        let account = list
            .iter()
            .find(|row| row.get("id").and_then(|v| v.as_i64()) == Some(account_id))
            .and_then(|row| row.get("name"))
            .and_then(|v| v.as_str())
            .ok_or_else(|| "请在日历清单中选择一个具体子日历".to_string())?
            .to_owned();
        let group = settings
            .get("group_id")
            .map(|v| v.1.clone())
            .unwrap_or_default();
        let unique = format!("dkcal_mdays_{}", day.format("%Y%m%d"));
        let rows: Vec<String> = {
            let mut stmt = db
                .prepare(
                    "SELECT it_content FROM item_table WHERE u_id=?1 AND pj_id=0 AND COALESCE(group_id,'')=?2 AND it_unique_id=?3",
                )
                .map_err(|e| e.to_string())?;
            stmt.query_map(rusqlite::params![account_id, group, unique], |row| row.get(0))
                .map_err(|e| e.to_string())?
                .collect::<Result<_, _>>()
                .map_err(|e| e.to_string())?
        };
        if rows.len() > 1 {
            return Err("当天日格存在多条记录，需先核对日历数据，未擅自合并".into());
        }
        let items = rows.first().map(|c| parse_items(c)).unwrap_or_default();
        let done = items
            .iter()
            .filter(|item| item.get("done") == Some(&json!(true)))
            .count();
        let weekday =
            ["一", "二", "三", "四", "五", "六", "日"][chrono::Datelike::weekday(&day).num_days_from_monday() as usize];
        let mut data = json!({
            "date": day.format("%Y-%m-%d").to_string(),
            "weekday": format!("周{weekday}"),
            "account": account,
            "account_id": account_id,
            "items": items.clone(),
            "total": items.len(),
            "done": done,
            "layout": 2,
        });
        let mut hasher = Sha256::new();
        hasher.update(json_dumps_sorted(&data).as_bytes());
        let signature = format!("{:x}", hasher.finalize())[..16].to_string();
        data.as_object_mut().unwrap().insert("signature".into(), json!(signature));
        data.as_object_mut()
            .unwrap()
            .insert("db_path".into(), json!(path.to_string_lossy()));
        // Connection drops at end of block.
        data
    };
    Ok(result)
}

pub fn render(snapshot: &Value) -> Result<Vec<Vec<u8>>, RenderError> {
    render::render_calendar_pages(snapshot)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::time::{SystemTime, UNIX_EPOCH};

    #[test]
    fn parse_completion_and_color_markup() {
        let items = parse_items(
            "普通[+]正文\r\n|&lt;|font color=|&quot;|#FF0000|&quot;||&gt;|[+]完成|&lt;|/font|&gt;|\r\n\r\n",
        );
        assert_eq!(items.len(), 2);
        assert_eq!(items[0]["text"], "普通[+]正文");
        assert_eq!(items[0]["done"], false);
        assert_eq!(items[1]["text"], "完成");
        assert_eq!(items[1]["done"], true);
    }

    #[test]
    fn collect_today_account_only() {
        let dir = std::env::temp_dir().join(format!(
            "holo-cal-{}",
            SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_nanos()
        ));
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("calendar.db");
        {
            let db = Connection::open(&path).unwrap();
            db.execute_batch(
                "CREATE TABLE setting_table(st_name TEXT,st_nval INTEGER,st_sval TEXT);
                 CREATE TABLE item_table(u_id INTEGER,pj_id INTEGER,group_id TEXT,it_unique_id TEXT,it_content TEXT);",
            )
            .unwrap();
            let accounts = r#"{"vdata":{"list":[{"id":1,"name":"工作"},{"id":2,"name":"生活"}]}}"#
                .replace('"', "|&quot;|");
            db.execute(
                "INSERT INTO setting_table VALUES(?1,?2,?3)",
                rusqlite::params!["sys_current_sub_account", 1i64, ""],
            )
            .unwrap();
            db.execute(
                "INSERT INTO setting_table VALUES(?1,?2,?3)",
                rusqlite::params!["sys_sub_account_list", 0i64, accounts],
            )
            .unwrap();
            db.execute(
                "INSERT INTO setting_table VALUES(?1,?2,?3)",
                rusqlite::params!["group_id", 0i64, ""],
            )
            .unwrap();
            db.execute(
                "INSERT INTO setting_table VALUES(?1,?2,?3)",
                rusqlite::params!["user_token", 0i64, "DO_NOT_READ"],
            )
            .unwrap();
            for (u, g, uid, content) in [
                (1i64, "", "dkcal_mdays_20260908", "待办\r\n[+]已完成"),
                (2, "", "dkcal_mdays_20260908", "别的子日历"),
                (1, "team", "dkcal_mdays_20260908", "别的团队"),
                (1, "", "dkcal_mdays_20260907", "不是当日"),
            ] {
                db.execute(
                    "INSERT INTO item_table VALUES(?1,0,?2,?3,?4)",
                    rusqlite::params![u, g, uid, content],
                )
                .unwrap();
            }
        }
        let before = std::fs::read(&path).unwrap();
        let day = chrono::NaiveDate::from_ymd_opt(2026, 9, 8).unwrap();
        let result = collect(Some(&path), Some(day)).unwrap();
        assert_eq!(result["total"], 2);
        assert_eq!(result["done"], 1);
        assert_eq!(result["account"], "工作");
        assert_eq!(result["items"][0]["text"], "待办");
        assert_eq!(result["items"][1]["done"], true);
        assert!(!serde_json::to_string(&result).unwrap().contains("DO_NOT_READ"));
        assert_eq!(std::fs::read(&path).unwrap(), before);
        let again = collect(Some(&path), Some(day)).unwrap();
        assert_eq!(result["signature"], again["signature"]);
        let empty = collect(
            Some(&path),
            Some(chrono::NaiveDate::from_ymd_opt(2026, 9, 9).unwrap()),
        )
        .unwrap();
        assert_eq!(empty["total"], 0);
        let _ = std::fs::remove_dir_all(&dir);
    }
}
